### PR TITLE
Fix Causality Handling in Flash Attention to Support Bidirectional Attention

### LINKED ISSUE #39554: Why `is_causal` is not used in `flash_attention_forward` ?
I want to perform bidirectional attention in the Qwen3 model to train an embedding model, so I passed `is_causal=False` in the model `forward` (I manually added `is_causal` arguments in all `forward` method such as `Qwen3Model` and `Qwen3Attention` in`modeling_qwen3.py`):

```python
class Qwen3Attention(nn.Module):
    """Multi-headed attention from 'Attention Is All You Need' paper"""
        ...

    def forward(
        self,
        hidden_states: torch.Tensor,
        position_embeddings: tuple[torch.Tensor, torch.Tensor],
        attention_mask: Optional[torch.Tensor],
        past_key_value: Optional[Cache] = None,
        cache_position: Optional[torch.LongTensor] = None,
        is_causal: Optional[bool] = True,   # I add is_causal here
        **kwargs: Unpack[FlashAttentionKwargs],
    ) -> tuple[torch.Tensor, Optional[torch.Tensor], Optional[tuple[torch.Tensor]]]:
        ...

        attn_output, attn_weights = attention_interface(
            self,
            query_states,
            key_states,
            value_states,
            attention_mask,
            dropout=0.0 if not self.training else self.attention_dropout,
            scaling=self.scaling,
            sliding_window=self.sliding_window,  # diff with Llama
            is_causal=is_causal, # and is_causal from the argument is passed to the attention_interface (e.g. `flash_attention_2`, `sdpa_attention_forward`)
            **kwargs,
        )
```
   
I can successfully change the causality of the attention in `sdpa_attention_forward`. However, I realized that it does not change the causality in the attention in `flash_attention_forward`. After diving into the implementation of `flash_attention_forward`, I found the reason in `flash_attention_forward` located at `transformers/integrations/flash_attention.py`:

```python
def flash_attention_forward(
    module: torch.nn.Module,
    query: torch.Tensor,
    key: torch.Tensor,
    value: torch.Tensor,
    attention_mask: Optional[torch.Tensor],
    dropout: float = 0.0,
    scaling: Optional[float] = None,
    sliding_window: Optional[int] = None,
    softcap: Optional[float] = None,
    **kwargs,
) -> tuple[torch.Tensor, None]:
    ...

    # FA2 always relies on the value set in the module, so remove it if present in kwargs to avoid passing it twice
    kwargs.pop("is_causal", None)

    attn_output = _flash_attention_forward(
        query,
        key,
        value,
        attention_mask,
        query_length=seq_len,
        is_causal=module.is_causal,  # here module is `Qwen3Attention`
        dropout=dropout,
        softmax_scale=scaling,
        sliding_window=sliding_window,
        softcap=softcap,
        use_top_left_mask=_use_top_left_mask,
        target_dtype=target_dtype,
        attn_implementation=module.config._attn_implementation,
        **kwargs,
    )
```

As you can see, the `is_causal` argument is popped, and the `is_causal` of `Qwen3Attention` is used as the argument. Note that `Qwen3Attention.is_causal` is never changed, and its default value is `True`, so the `is_causal` argument passed into `_flash_attention_forward` will always be `True` regardless of any change. 

I would like to know if it is a feature or a bug? Thank you!!

### PR DESCRIPTION
**Pull request by lucaswychan**
# What does this PR do?
Fixes #39554 

The original implementation of the `flash_attention_forward` function is restricted to performing **causal attention** and does not support **bidirectional attention**. This behavior stems from how the function handles causality:

- The function relies on the `Attention.is_causal` attribute, which belongs to the `Attention` class in the model.
- By default, `Attention.is_causal` is set to `True`, enforcing causal attention (where the model only attends to previous tokens in a sequence).
- This attribute is never modified in the code, meaning the setting is effectively fixed.
- Additionally, while the function removes the `is_causal` key from the keyword arguments (`kwargs`) passed to it, this value is not used. Instead, it always defers to the hardcoded `Attention.is_causal` value.

As a result, even if a user attempts to pass `is_causal=False` through `kwargs` to enable bidirectional attention (where the model can attend to both previous and future tokens), the input is ignored. Consequently, the current setup makes it impossible to perform bidirectional attention when using flash attention.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker @Cyrilvallez @vasqu

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
BTW checked this: #35390 let's make sure we don't break it!

**Comment by lucaswychan** [comment]
> BTW checked this: #35390 let's make sure we don't break it!

**Comment by lucaswychan** [comment]
After checking #35390 , it seems like it only pops `is_causal` from `kwargs` without saving and using it (which is the reason I created this PR). My approach takes a similar way, but I first store the value of `is_causal` when popping, and fallback to `module.is_causal` only when `is_causal is None` (which means `is_causal` does not exist in `kwargs`). Below are the sample code to check my fix:

```python
import torch
from transformers import AutoTokenizer, AutoModel

model_name = "Qwen/Qwen2.5-7B"
device = torch.device("cuda")

tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModel.from_pretrained(model_name, device_map=device, attn_implementation="flash_attention_2", torch_dtype="auto")

input_sentence = "I want bidirectional attention"
inputs = tokenizer([input_sentence], return_tensors="pt", padding=True).to(model.device)

output_with_causal = model(**inputs, is_causal=True)
print(f'last hidden state with causal : {output_with_causal["last_hidden_state"]}')

output_without_causal = model(**inputs, is_causal=False)
print(f'last hidden state without causal : {output_without_causal["last_hidden_state"]}')
```

and you should see the output:

```
last hidden state with causal : tensor([[[-2.1719,  2.5469, -6.6875,  ...,  2.9531, -0.2100,  2.2031],
         [-0.7305, -0.2656, -1.5156,  ..., -0.7461, -0.0952, -1.6641],
         [-0.4668,  1.8125,  0.2021,  ...,  0.8867, -1.1875, -1.0547],
         [ 1.8359,  2.3750, -1.5547,  ...,  0.6445, -1.3906, -2.0469],
         [ 4.1562, -3.6250, -2.9844,  ..., -2.1562, -1.2031, -0.3086]]],
       device='cuda:7', dtype=torch.bfloat16, grad_fn=<MulBackward0>)

last hidden state without causal : tensor([[[ 3.0156, -0.7070, -1.5234,  ..., -1.1875,  0.7188, -0.0464],
         [ 0.4629, -1.0781,  0.0135,  ..., -2.2656,  0.2412, -2.6250],
         [ 0.2334, -2.0469, -0.6680,  ..., -1.8906, -0.7305, -3.2969],
         [ 0.9961, -0.9844, -0.9805,  ..., -1.2500, -0.3555, -2.1094],
         [ 1.3125,  0.7539, -0.5781,  ..., -0.4824,  0.0216,  0.3633]]],
       device='cuda:7', dtype=torch.bfloat16, grad_fn=<MulBackward0>)
```

which indeed the output is changed by altering `is_causal`.

**Review by ArthurZucker — APPROVED** [review]
It is actually a nice feature to have! 
Pinging @vasqu to make sure we are not forgetting anything!

**Comment by ArthurZucker** [comment]
run-slow: auto

**Comment by ArthurZucker** [comment]
https://github.com/huggingface/transformers/issues/39554#issuecomment-3133120685 as per his comment we need to make sure the attention mask creation takes this into account imo!

**Comment by lucaswychan** [comment]
I believe the attention mask creation is handled well in FA2, as reflected by the function `flash_attn_varlen_func`, which allows passing an argument `causal` (the same as `is_causal` with a different name) to control the causality of the attention. 

For SDPA, I ran a small experiment to see if we can enable bidirectional attention, and the answer is yes evidenced by the following result:
```
# SDPA

last hidden state with causal : tensor([[[-2.1719,  2.5469, -6.6875,  ...,  2.9531, -0.2100,  2.2031],
         [-0.7227, -0.2656, -1.5156,  ..., -0.7383, -0.0923, -1.6719],
         [-0.4590,  1.8281,  0.1631,  ...,  0.9141, -1.1875, -1.0859],
         [ 1.8359,  2.3594, -1.5625,  ...,  0.6484, -1.3516, -2.0625],
         [ 4.1875, -3.6406, -3.0000,  ..., -2.1562, -1.2578, -0.2637]]],
       device='cuda:6', dtype=torch.bfloat16, grad_fn=<MulBackward0>)
last hidden state without causal : tensor([[[ 2.8906, -0.6758, -1.5234,  ..., -1.1797,  0.7148, -0.0840],
         [ 0.4395, -1.1094,  0.0476,  ..., -2.3281,  0.1611, -2.6094],
         [ 0.2158, -2.1562, -0.6406,  ..., -1.9297, -0.7266, -3.3438],
         [ 0.9023, -1.1484, -1.0000,  ..., -1.3828, -0.3301, -2.1719],
         [ 1.2500,  0.5391, -0.4492,  ..., -0.4629,  0.0056,  0.2520]]],
       device='cuda:6', dtype=torch.bfloat16, grad_fn=<MulBackward0>)
```

**Review by vasqu — APPROVED** [review]
It should be fine, fa is the "easy" case :D see my comment at https://github.com/huggingface/transformers/issues/39554#issuecomment-3178976605 

I'm still pro this, especially since this is a power feature for now. In the future, it would be nice to enable this for other flavors (sdpa, eager, flex) as well - this will need more adjustments tho (mask).

Not entirely sure what you did with SDPA there, but the `causal` flag in sdpa is a mess¹ tbh and having bidirectionality vs causality is not as easy when padding is involved (which is when we have to rely on the mask ~ pretty much all the time).

¹ the flag is only allowed without an attention mask and depending on the query length sometimes we need to disable it as well even with causal attention (during inference)

**Comment by vasqu** [comment]
Can you merge/rebase? Just need to make sure CI is green

**Comment by vasqu** [comment]
Thx for the PR @lucaswychan :hugs: (and for bearing with me :p)

**Comment by PavloFesenko** [comment]
@lucaswychan One tiny nitpick: could you please update `Refer to #<issue number>` to `Fixes #<issue number>` in the PR description? The word `Fixes` tells GitHub that this PR will fix the underlying issue, displays a PR icon near the issue in the issue list, and automatically closes the issue once the PR is merged. I don't know if it will close the issue now since the PR is already merged but you can always close the underlying issue manually. 🙏

**Comment by Inistlwq** [comment]
@vasqu hello，can you share how i can convert qwen3 to bidirectional attention for embedding detailly thank u

**Comment by vasqu** [comment]
@Inistlwq You can just pass `is_causal=False` in your forward, e.g. something along 
```
...
output = model.forward(**inputs, is_causal=False)
...
```
Note that this will only work with Flash Attention out of the box.

**Comment by Inistlwq** [comment]
@vasqu  I tested it on v100，which doesn't support Flash Attention，I tried pass is_causal=False many times，but it doesn't work。If no Flash Attention，how I can convert qwen3 to bidirectional attention

**Comment by Inistlwq** [comment]
@vasqu `batch_dict = tokenizer(
    input_texts,
    padding=True,
    truncation=True,
    max_length=max_length,
    return_tensors="pt",
)
batch_dict.to(model.device)
outputs = model(**batch_dict,is_causal=False)` and i just test it on h20，it does work ；but if don't want to use flash attention,how I can convert ,for example on v100

**Comment by vasqu** [comment]
@Inistlwq

For Flash Attention, you can use https://github.com/huggingface/transformers/pull/39707#issuecomment-3127851175
```python
import torch
from transformers import AutoTokenizer, AutoModel

model_name = "Qwen/Qwen2.5-7B"
device = torch.device("cuda")

tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModel.from_pretrained(model_name, device_map=device, attn_implementation="flash_attention_2", torch_dtype="auto")

input_sentence = "I want bidirectional attention"
inputs = tokenizer([input_sentence], return_tensors="pt", padding=True).to(model.device)

output_with_causal = model(**inputs, is_causal=True)
print(f'last hidden state with causal : {output_with_causal["last_hidden_state"]}')

output_without_causal = model(**inputs, is_causal=False)
print(f'last hidden state without causal : {output_without_causal["last_hidden_state"]}')
```
For anything else, you will need to provide your own mask
```python
import torch
from transformers import AutoTokenizer, AutoModel
from transformers.masking_utils import create_bidirectional_mask


model_name = "Qwen/Qwen2.5-7B"
device = torch.device("cuda")

tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModel.from_pretrained(model_name, device_map=device, torch_dtype="auto")

input_sentence = "I want bidirectional attention"
inputs = tokenizer([input_sentence], return_tensors="pt", padding=True).to(model.device)

output_with_causal = model(**inputs)
print(f'last hidden state with causal : {output_with_causal["last_hidden_state"]}')

inputs["attention_mask"] = create_bidirectional_mask(
    config=model.config,
    input_embeds=torch.ones_like(inputs["input_ids"])[..., None].to(model.dtype),
    attention_mask=inputs["attention_mask"],
)
output_without_causal = model(**inputs, is_causal=False)
print(f'last hidden state without causal : {output_without_causal["last_hidden_state"]}')
```
Note, that flash attention does not really need extra care for the mask while the other variations do.

**Comment by Inistlwq** [comment]
> ```python
> create_bidirectional_mask
> ```

@vasqu  thank u very much，I will try!
